pipenv/vendor/requirementslib/models/requirements.py

```diff
@@ -2568,7 +2568,8 @@ def from_pipfile(cls, name, pipfile):
                 r.req.marker = req_markers.marker
         extras = _pipfile.get("extras")
         if r.req:
-            r.req.specifier = SpecifierSet(_pipfile["version"])
+            if r.req.specifier:
```

Comment:
I am not sure I understand this -- if the specifier is already defined, maybe we shouldn't be over-writing it, but I think if it isn't defined, we want to add the Specifier set -- but we need to somehow detect if there is not a specifier in front of a version number that we intend for that to be `==`.

Reply:
This change is the revert of https://github.com/pypa/pipenv/commit/7458b32e345db109bc94f03ce50a3c6e15c99185#diff-343ecee9e3cdaa822811a0c53132eaed220298797db5a50b019f17d404d3184cL2763 if that can give you context or recall you anything. 

I am not familiar on all the concept, `r.req` is  `Requirement.parse('fastapi0.95.0')` and `r.req.specifier` is `<SpecifierSet('')>` before overriding. I am not sur where the "default operator" information should belong. 

Passing something called version but somehow expecting an operator is confusing. However I can see that this variable carry the operator versoin when the Pipfile contains `==X.Y.X`. 
Should we update the get_version to ensure at least an operator is defined ? Either that or get_operator method

```python
_pipfile["operator"] = get_operator(pipefile)
SpecifierSet("{}{}".format(_pipfile["operator"],_pipfile["version"])

def get_operator(pipfile_entry):
    # similar logic than get_version, try to get various key and defaulting to "=="
    # also need to update get_version to be operator-aware to discard that part
```

Either that or a more pythonic way : 
```python
try:
    r.req.specifier = SpecifierSet(_pipfile["version"])
except InvalidSpecifier:
    r.req.specifier = SpecifierSet("=={}".format(_pipfile["version"]))
```
That would still raise `InvalidSpecifier` with a bad input.

Reply:
> if that can give you context or recall you anything.

Oh I recall much about the 40 hours I dumped into refactoring that and trying to unravel some of it, but what I learned is there is still a long ways to go before requirementslib is "good code".   Anyway, the path is probably not to revert but to fix forward.

> r.req is Requirement.parse('fastapi0.95.0')

That feels wrong I don't think Requirement.parse should handle that because there is no operator between the name and the specifier.

I like and think we should head in the direction you outlined:

```
_pipfile["operator"] = get_operator(pipefile)
SpecifierSet("{}{}".format(_pipfile["operator"],_pipfile["version"])

def get_operator(pipfile_entry):
    # similar logic than get_version, try to get various key and defaulting to "=="
    # also need to update get_version to be operator-aware to discard that part
```